manifests/charts/istio-control/istio-discovery/files/gateway-injection-template.yaml

```diff
@@ -4,7 +4,7 @@ metadata:
   labels:
     service.istio.io/canonical-name: {{ index .ObjectMeta.Labels `service.istio.io/canonical-name` | default (index .ObjectMeta.Labels `app.kubernetes.io/name`) | default (index .ObjectMeta.Labels `app`) | default .DeploymentMeta.Name  | quote }}
     service.istio.io/canonical-revision: {{ index .ObjectMeta.Labels `service.istio.io/canonical-revision` | default (index .ObjectMeta.Labels `app.kubernetes.io/version`) | default (index .ObjectMeta.Labels `version`) | default "latest"  | quote }}
-    istio.io/rev: {{ .Revision | default "default" | quote }}
+    istio.io/rev: {{ index .ObjectMeta.Labels `istio.io/rev` | default .Revision | default "default" | quote }}
```

Comment:
Hmm so the difference between this and the original implementation was before we injected the **actual revision** and now we inject the **desired revision** (which could be a tag).

I like it in some ways (it at least solves the original issue), but don't like it in other ways.

For example, if I have revisions v1 and v2 with tag "canary" and "stable", I may think I can do `kubectl get pods -A -l istio.io/rev=v1` (or look at metrics, since I think you said that was your use case) to find all v1 pods. But this is actually not the case! if a user had `istio.io/rev=canary` which resolves to v1, the pod will also have `istio.io/rev=canary` and we will miss it.

This is not necessarily a bug if we consider the label to mean "desired revision" but I am a bit worried people will assume it is "actual revision"

WDYT @linsun @Monkeyanator

Reply:
> before we injected the actual revision and now we inject the desired revision (which could be a tag).

Yes, you're right, but it doesn't matter from the injection point of view.

I know that users cannot get pods by revision when they use revision tags, but this is the drawback of revision tags, which actually are not required. For me, this is a little downside comparing to consequences of creating pods until resources are exhausted.

Reply:
> Hmm so the difference between this and the original implementation was before we injected the actual revision and now we inject the desired revision (which could be a tag).

Just for clarification: this is necessary to fix #33237, because otherwise tag and revision will never match and it will cause infinite creating pods.

Reply:
Yep I get that. I think the tradeoff is no label vs possibly confusing label, not regressing on https://github.com/istio/istio/issues/33237 (which would be bad). I am on the fence, will look for others opinions

Reply:
No, the trade-off is gateway injection works with tag revisions vs confusing label.

Reply:
The change in sidecar injector is not relevant for the bug fix. I should made 2 different PRs.

Reply:
I would think the two should be consistent though, right? So if we align with sidecar injector we should remove the label entirely?

Reply:
And they are consistent right now. No, my goal is to have this label on all workloads.

Reply:
Sorry I don't get how its consistent with this PR now?

Gateway: inject istio.io/rev label
Sidecars: do not inject istio.io/rev label?

The initial version was consistent since you *also* added it for sidecars, but its not anymore...?

I think the two need to be consistent. I am probably fine with it consistently being "inject istio.io/rev label, with logic here" but i want more feedback from other folks before unilaterally making a change that has wide implications.

Reply:
No, currently gateways have `istio.io/rev` and sidecars don't have. Sidecars had this label in the past, but it was removed in #33447 as a fix for #33237, but in fact it didn't fix the issue with gateway injection. So this change aims to make gateways and sidecars consistent and inject `istio.io/rev` in both cases and at the same time, solve the problem of gateway injection when revision tag is specified.

Reply:
I thought you said you were reverting the change to manifests/charts/istio-control/istio-discovery/files/injection-template.yaml. Sorry for the miscommunication.

In that case:

> I am probably fine with it consistently being "inject istio.io/rev label, with logic here" but i want more feedback from other folks before unilaterally making a change that has wide implications.

Reply:
Because I reverted that change to make this consistently. Anyway, if you're fine with this, let's wait for feedback from others.

Reply:
Q: if canary is changed to v3, do we go back to change the pod to be labelled as v3? Or would that only happen if the pod gets restarted?

I think I am leaning towards to be very explicit about the revision for gateways, which would be v2 if that is the revision. Tag is introduced as a convenience for apps, for gateways, I can see users want to be more explicit as it could impact many apps?

Reply:
After thinking about this more I am a bit worried about the inconsistency. Let's consider we have a revision R and tag T

If namespace has R, pod gets R
If pod has R, pod gets R
If namespace has T, pod gets R
If pod has T, pod gets T

Reply:
> if canary is changed to v3, do we go back to change the pod to be labelled as v3? Or would that only happen if the pod gets restarted?

If you remove tag `canary` and then create `v3`, a gateway will work until you restart its pod. But a sidecar would behave identically.

> I think I am leaning towards to be very explicit about the revision for gateways

Do you suggest to ignore injection when a gateway is labeled with a tag instead of a revision? What about sidecars then? Sidecars currently work fine when are labeled with `istio.io/rev=<tag>`.

Reply:
The list of inconsistent labeling is valid only for gateways, so in fact currently we have the following rules:
1. Gateways:
- If namespace has R, pod gets R
- If pod has R, pod gets R
- If namespace has T, pod gets R
- If pod has T, pod get R **(and therefore pods fail!)**
2. Sidecars:
- If namespace has R, pod does not get label
- If pod has R, pod gets R
- If namespace has T, pod does not get label
- If pod has T, pod gets T

This PR will change labeling rule as follows:
1. Gateways:
- If namespace has R, pod gets R
- If pod has R, pod gets R
- If namespace has T, pod gets R
- If pod has T, pod get T
2. Sidecars:
- If namespace has R, pod gets R
- If pod has R, pod gets R
- If namespace has T, pod gets R
- If pod has T, pod gets T

So as you can see, this PR makes labeling more consistent that this is for now.

Reply:
Right, my suggestion is to make gateways like sidecars are today and not mess with any label. Then in all 4 categories the result is "we don't mess with the label".

Reply:
Making gateways like sidecars are today means that gateway injection supports "If pod has T, pod gets T" as sidecars support it. So that's great, because this is my goal.

If it comes to making gateways like sidecars in case of "If namespace has R/T, pod does not get label", we can't do this, because it would be backward incompatible. Btw. propagating namespace labels is not related to this PR.

If something is really wrong with this change, please point what it is, because I completely don't understand your position to this PR. I solved a bug which looks major or critical, because it directly affects performance of kube-apiserver and etcd and you block this, because you worry what labeling rules users may expect while we even don't have data to draw conclusions. In fact, we just suppose while it may not be a problem at all, and at the same time, it makes collecting metrics per revision more complicated. I don't understand why do you ignore this goal.

Reply:
> The only way I see to achieve those requirements and meet your goal is to remove the injected istio.io/rev label on gateway pods (no change needed to sidecar as it was already removed).

Removing injection of `istio.io/rev` to pods prevents collecting metrics per revision, so what do you think about changing case "If namespace has T, pod gets R" to "If namespace has T, pod gets T"?

Reply:
"If namespace has T, pod gets T"?

How will this be implemented? We don't have access to this info currently in the injection

I still don't think it meets your goal of "metrics per revision" since it's *not the revision*, it's the revision OR the tag which makes it not useful for most purposes

Reply:
I took a look how `revision tag set` works and I guess that it should be feasible to create webhook so that it knows what should be injected to istio.io/rev. If setting revision based on tag is possible I guess that setting tag should be as well.

> I still don't think it meets your goal of "metrics per revision"

This is true, but I don't want to remove the rule "If namespace has R, pod gets R" and therefore I care about the case with tag. It's not so simple to keep "If namespace has R, pod gets R" and removing "If namespace has T, pod gets R". This would require an additional refactor in injection webhooks.

I think that the best now is to narrow this PR to just fix gateway injection issue and I will address the problem of propagating namespace label to pods in a subsequent PR in which I will try to implement "If namespace has T, pod gets T".

Reply:
Just to clarify, I am not sure "If namespace has T, pod gets T" is appropriate either. It's highly confusing to have a label injected, for purpose of metrics or identifying revision usage etc when they it's NOT actually recording the revision. 

IMO there are 2 valid paths:
* remove the istio.io/rev label injection in gateway (align with sidecar)
* Make a new label (strawman: istio.io/actual-revision). Always set this to .Revision. Reject pods that have this label already set to avoid the Deployment selector issue (except.. this may break idempotency and not be viable...). This is like the old logic but attempting to avoid the Deployment issue

The latter would be a new API and I'm not sure it's a good idea at this point (nor am I sure it actually avoids the selector issue safely). The former I will certainly approve, although I don't think it meets your use cases(?).

Reply:
As mentioned above I am not okay with the change here. See the options in https://github.com/istio/istio/pull/43433/files#r1112378489.

To be clear - I acknowledge you are fixing a bug here. However, "fixes a bug" is not the only criteria to accept a PR. This introduces a regression in a variety of ways (understability, semantics, consistency).

How about we just change this to remove the istio.io/rev label injection?

Reply:
Please look at this once again. Currently, label injection works as follows:

Gateways:
1. If namespace has R, pod gets R
2. If pod has R, pod gets R
3. If namespace has T, pod gets R
4. **If pod has T, pod get R (injection fails!)**

Sidecars:
1. If namespace has R, pod does not get label
2. If pod has R, pod gets R
3. If namespace has T, pod does not get label
4. If pod has T, pod gets T

After reverting changes made in sidecar injection template, labeling will work as follows:

Gateways:
1. If namespace has R, pod gets R
2. If pod has R, pod gets R
3. If namespace has T, pod gets R
4. **If pod has T, pod get T**

Sidecars: **(rules didn't change**)
1. If namespace has R, pod does not get label
2. If pod has R, pod gets R
3. If namespace has T, pod does not get label
4. If pod has T, pod gets T

As you can see, I changed only the failing scenario, which is consistent with current sidecar injection and I didn't introduce any regression, understability, etc.

> How about we just change this to remove the istio.io/rev label injection?

As I already said, I really want keep case "If namespace has R, pod gets R", because we will really need this to support multi-revision observability. I know that you suggested alternatives, but those also require analysis of pros and cons, and I am not sure if we really need to remove `istio.io/rev` in favor of another label. I don't want to change this PR back and forth. I will address that problem in a follow-up PR and I will take into account your expectations.

Reply:
When you phrase it like that (which, admittedly, was how I phrased it :-) ), I agree it makes Sidecar and Gateway the same with this change. however, the missing case is

"If pod has no label"

Gateways with this change: "If pod has no istio.io/rev label, gets R"
Sidecars: "If pod has no istio.io/rev label, gets no label"

That is the inconsistency... the `istio.io/rev` has different meanings at different times.

The sidecar case as today could be simplified to "The pod label is not mutated". The gateway case cannot.

---

Its not clear to me why you need multi-revision observability *only* for gateways..? Is that actually the case? If not, why do we want them to behave differently.

Reply:
But I don't try to address case "If pod has no label". The point is that "If pod has label T, gateway injection fails!", and current scope of this PR addresses this case. Nothing else, because other cases work. I did a mistake trying to address to different things in a single PR, but I already limited scope and updated title.

If it comes to `Gateways with this change: "If pod has no istio.io/rev label, gets R"` - this is implied by "If namespace has R/T, pod get R/T". But please note that I didn't add this - I just didn't remove this, because I want to fix the bug keeping backward compatibility.

> Its not clear to me why you need multi-revision observability only for gateways..?

We need multi-revision observability for gateways as well as for sidecars. And this is why I initially reverted changes made in the #33447, which removed the rule "If namespace has R, pod gets R", but then reverted this from my PR, because of your objections. Removing istio.io/rev entirely, removes the rule "If namespace has R, pod gets R" which we really need for multi-revision observability. Otherwise, this will be much less usable.

Reply:
> But I don't try to address case "If pod has no label"

And I am saying we cannot ignore these cases...

We are going in circles, lets discuss in the WG meeting.